### ISSUE TITLE
Show class names in SHAP summary plots

### ISSUE THREAD

**Issue by Mark531** [shap/shap#764]
Hello,

For a reason I ignore, SHAP summary plots don't show class names by default:

![image](https://user-images.githubusercontent.com/17707763/63474855-e7788400-c47a-11e9-8571-1995a3900499.png)

The default names can be changed by using the class_names parameter, but how can I know which class number is what?

Can't SHAP know the right names of the classes from the model?

Mark

**Comment by slundberg** [comment]
Most model objects don't have names for their outputs, they just have indexes, so that is what get's shown by default. So if the model has 4 dimensional outputs the class labels are the indexes of each output. Hope that helps.

**Comment by Mark531** [comment]
Your remark is right for a neural network but actually my model is a multiclass XGBoost model, sorry I forgot to mention it. So, it has only one output, but 4 trained models in memory. The classes should be known by the model, because they're passed in the y_train/y_val parameters. Since the classes have no particular order, I can't know from your graph which class corresponds to what.

**Comment by slundberg** [comment]
Ah, so if you call predict do you a categorical output? If you do predict_proba (with the sklearn api) you should get a matrix that matches the order given by shap. If you have a specific example I can also look

**Comment by Mark531** [comment]
Of course, predict() return classes, it's the contract met by all sklearn's estimators.

Here is a minimal example reproducing the problem:
```
import numpy as np
from xgboost import XGBClassifier
import shap

X = np.array([[1, 2], [3, 4], [5, 6], [7, 8], [9, 10]])
y = np.array(["A", "B", "C", "A", "B"])

model = XGBClassifier()
model.fit(X, y, eval_metric="mlogloss")
y_pred = model.predict(X)

explainer = shap.TreeExplainer(model)
shap_values = explainer.shap_values(X)
shap.summary_plot(shap_values, X, plot_type="bar")
```
In the SHAP summary plot, you'll see "Class 0", "Class 1" and "Class 2" instead of "A", "B" and "C". I think it could be improved, especially as the classes' names are stored in the model (field classes_). Currently, in this form, the SHAP summary plot cannot be exploited.

**Comment by slundberg** [comment]
Right. Currently the indexes used are the same as those used by the model (such as with `model.predict_proba(X)`). It would indeed be nice if the summary plot could automatically know or attach names to the outputs. The problem is then we would need to return not a standard `list` of SHAP values (one for each output), but some custom class that could also hold the names of the outputs. That could bring with it a lot of complications, since dealing with custom objects is more tricky. We might want to do this in the future but it would be an API change.

For now you can (as you mentioned) just use the `class_names=model.classes_` option of the summary plot.

**Comment by slundberg** [comment]
I added a todo label to mark the idea of creating a custom class for future consideration.

**Comment by Mark531** [comment]
Oh you're right, I didn't pay attention to the fact that the summary plot only had access to the SHAP values, not to the model. So, I understand that my request would incur an API change. Thanks for your analysis.

**Comment by 610yilingliu** [comment]
For those who googled this issue:
Solution: add class_names in shap.summary_plot
example using titanic dataset:

```
shap.summary_plot(shap_values = shap_matrix,features = x, feature_names=x.columns.tolist(),  plot_type='bar', class_names=['Died', 'Survived']) #Died is value 0 in Survived column, Survived is value 1 in Survived column
```
![image](https://user-images.githubusercontent.com/41269216/166214784-3646e644-f3b3-4f7a-8af6-2070fa494dd7.png)

**Comment by ShuyangenFrance** [comment]
I would like to create a pull request for this issue.

**Comment by arilwan** [comment]
I map `class_inds` to a dict of `class_names` like so:
```python
# ['setosa', 'versicolor', 'virginica']
shap.summary_plot(shap_values, X_test, class_names={0:'setosa', 1:'versicolor', 2:'virginica'},
 class_inds='original')
```
![Screenshot from 2023-07-03 17-16-10](https://github.com/slundberg/shap/assets/42409395/738a657b-f16a-47fd-aed4-8b860f8a8293)
